lex/util/lex_format.go

```diff
@@ -0,0 +1,592 @@
+package util
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"regexp"
+
+	"github.com/ipfs/go-cid"
+	"github.com/relvacode/iso8601"
+	"golang.org/x/xerrors"
+
+	cbg "github.com/whyrusleeping/cbor-gen"
+)
+
+// cbor-gen internally uses reflect to determine the type of the field, and
+// if the type definition is an alias, it will treat the type as the underlying type(string) instead of the new name.
+// For this reason, we need to define a new struct for each format types, not using an alias.
```

Comment:
I may be the root of this confusion, speaking go with a rust accent or using incorrect terminology ("alias" vs "wrapper type"?). please do push back / correct on how to do this most idiomatically in golang.

naive notes on implementation options my linked issue: https://github.com/bluesky-social/indigo/issues/101

existing similar situations for `[]bytes` and `cid.Cid` in lexgen:

* https://github.com/bluesky-social/indigo/blob/main/lex/util/lex_types.go#L86
* https://github.com/bluesky-social/indigo/blob/main/lex/util/lex_types.go#L21

Reply:
Ah, sorry for the confusion! I used the term "alias" for the wrapped type: `type WrappedType UnderlyingType,` similar to rust newtype.

The reason the wrapped type didn't work for string is that when [cbor-gen](https://github.com/whyrusleeping/cbor-gen/blob/master/gen.go#L1395) tries to generate code from the schema-generated struct fields, the `reflect.Type()` in the switch statement uses the underlying type instead of the wrapped type itself. `type LexLink cid.Cid works okay because the underlying type is a struct instead of a string, so it properly generates a code that invokes `MarshalCBOR` and `UnmarshalCBOR.` 

When the underlying type is a string, it simply generates the code with `emitCborUnmarshalStringField.` here is a piece of code when I lexgen'ed from string-wrapped identifier types:

```go
			// t.Repo (util.FormatDID) (string)
		case "repo":

			{
				sval, err := cbg.ReadString(cr)
				if err != nil {
					return err
				}

				t.Repo = util.FormatDID(sval)
			}
```

As you can see, the custom marshaler logic(with validation) is ignored, and it got read as a string.

Reply:
> Another way we can solve this would be adding an interface implementation checking logic(such as f.Type.Implements(reflect.TypeOf((*cbor.Marshaler)(nil)).Elem())) before the switch logic, and if the type have some custom marshaler, generate the code using it. However this requires upstream change on the cbor-gen repo.

I'm not a stakeholder here, but yeah, personally I'd say we should fix cbor-gen with something like that instead. It'd be much more ergonomic for users to have the underlying types be strings (so they can write untyped string constants) rather than make all callers have to use constructors to get the string types.

Reply:
cc @whyrusleeping who wrote that cbor-gen.

Reply:
If @whyrusleeping sees this is good to go, I can write a quick patch for this interface checking logic(plus some tests)!

Reply:
Hey @isofrequency really sorry for letting this one sit so long, I really like this approach, its quite clean. and im happy to have you help fix up the codegen too, im very un-proud of it